Kernel/semaphores.c

```diff
@@ -0,0 +1,163 @@
+            scr_print("Id: ");
+            scr_printDec(sem);
+            scr_print(", \t Value: ");
+            scr_printDec(semaphores[sem]->value);
+            scr_print(", \t Linked processes: ");
+            scr_printDec(semaphores[sem]->linkedProcesses);
+            scr_print(", \t Name: ");
+            scr_print(semaphores[sem]->name);
+            scr_printChar('\n');
+        }
+
+    }
+}
+
+int sem_init() {
+    namer = rnm_new();
+    generalLock = 0;
+    if(namer != 0)
+        return SEM_FAILED;
+    return 0;
+}
+
+Tsem sem_open(const char * name, uint8_t initialValue) {
+
+    _spin_lock(&generalLock);
+
+    TSemaphore * sem = rnm_getResource(namer, name);
+
+    if(sem!=NULL){
+        sem->linkedProcesses+=1;
+        generalLock = 0;
+        return sem->semId;
+    }
+    
+    int i;
+    for( i=0 ; i < MAX_SEMAPHORES && semaphores[i]; ++i );
+
+    if(i == MAX_SEMAPHORES) {
+        generalLock = 0;
+        return SEM_FAILED;
+    }        
+    
+    semaphores[i] = mm_malloc(sizeof(TSemaphore));
+    if(semaphores[i] == NULL){
+        generalLock = 0;
+        return SEM_FAILED;
+    }
+    semaphores[i]->value = initialValue;
+    semaphores[i]->semId = i;
+    semaphores[i]->lock = 0;
+    semaphores[i]->linkedProcesses = 1;
+    semaphores[i]->waitingProcesses = wq_new();
+
+    if(semaphores[i]->waitingProcesses == NULL){
+        mm_free(semaphores[i]);
+        generalLock = 0;
+        return SEM_FAILED;
+    }
+
+    if(rnm_nameResource(namer, semaphores[i], name, &(semaphores[i]->name))!=0){
+        //check what should be freed
+        generalLock = 0;
+        return SEM_FAILED;
+    }
+
+    generalLock = 0;
+    return (Tsem) i;
+}
+
+int sem_close(Tsem sem) {
+
+    _spin_lock(&generalLock);
+    
+    if(!isValidSemId(sem)){
+        generalLock = 0;
+        return SEM_NOTEXISTS;
+    }
+
+    _spin_lock(&(semaphores[sem]->lock));
+    generalLock = 0;    
+
+    if(semaphores[sem]->linkedProcesses == 1){
+        return sem_free(sem); //The semaphore is destroyed, so we do not care about its lock
+    }                       
+    
+    semaphores[sem]->linkedProcesses-= 1;
+    semaphores[sem]->lock = 0;
+    return SEM_SUCCES;
+}
+
+int sem_post(Tsem sem) {
+    if(!isValidSemId(sem))
+        return SEM_NOTEXISTS;
+    
+    if(semaphores[sem]->value > 0 || wq_unblockSingle(semaphores[sem]->waitingProcesses) != 0){
+        semaphores[sem]->value+=1;
+    }
+    return SEM_SUCCES;
+}
+
+int sem_wait(Tsem sem) {
+    if(!isValidSemId(sem))
+        return SEM_NOTEXISTS;
+    
+    if(semaphores[sem]->value > 0){
+        semaphores[sem]->value-=1;
+        return SEM_SUCCES;
+    }
+
+    TPid cpid = sch_getCurrentPID();
+
+    scr_printDec(cpid);
+    scr_printDec(1);
+    if((wq_addIfNotExists(semaphores[sem]->waitingProcesses, cpid ))==0) {
+        scr_printDec(cpid);
+        scr_printDec(2);
+        sch_blockProcess(cpid);
+        sch_yieldProcess();
+    }
```

Comment:
In line with my previous comment, the code for this should look something like this:
```c
TPid pid = sch_getCurrentPid();
while (semaphore->value == 0) {
    wq_add(semaphore->waitingProcesses, pid);
    sch_blockProcess(pid);
    sch_yieldProcess();
}

semaphore->value--;
return SEM_SUCCESS;
```